pkg/network/tracer/tracer_linux_test.go

```diff
@@ -1669,3 +1683,386 @@ func TestKprobeAttachWithKprobeEvents(t *testing.T) {
+				cancel()
+				require.NoError(t, err)
+				defer c.Close()
+				c.Write([]byte("HTTP/1.1 404 NOT FOUND\r\n"))
+			},
+			serverRun: func(t *testing.T, serverAddr string, done chan struct{}) {
+				server := NewTCPServerOnAddress(serverAddr, func(c net.Conn) {
+					r := bufio.NewReader(c)
+					input, err := r.ReadBytes(byte('\n'))
+					if err == nil {
+						c.Write(input)
+					}
+					c.Close()
+				})
+				require.NoError(t, server.Run(done))
+			},
+			want: network.ProtocolHTTP,
+		},
+		{
+			name: "gRPC traffic - unary call",
+			done: make(chan struct{}),
+			clientRun: func(t *testing.T, serverAddr string) {
+				c, err := grpc.NewClient(serverAddr, grpc.Options{
+					CustomDialer: dialer,
+				})
+				require.NoError(t, err)
+				defer c.Close()
+				require.NoError(t, c.HandleUnary(context.Background(), "test"))
+			},
+			serverRun: func(t *testing.T, serverAddr string, done chan struct{}) {
+				closer, err := grpc.Server(serverAddr)
+				require.NoError(t, err)
+				go func() {
+					<-done
+					closer()
+				}()
+			},
+			want: network.ProtocolHTTP2,
+		},
+		{
+			name: "gRPC traffic - stream call",
+			done: make(chan struct{}),
+			clientRun: func(t *testing.T, serverAddr string) {
+				c, err := grpc.NewClient(serverAddr, grpc.Options{
+					CustomDialer: dialer,
+				})
+				require.NoError(t, err)
+				defer c.Close()
+				require.NoError(t, c.HandleStream(context.Background(), 5))
+			},
+			serverRun: func(t *testing.T, serverAddr string, done chan struct{}) {
+				closer, err := grpc.Server(serverAddr)
+				require.NoError(t, err)
+				go func() {
+					<-done
+					closer()
+				}()
+			},
+			want: network.ProtocolHTTP2,
+		},
+		{
+			// A case where we see multiple protocols on the same socket. In that case, we expect to classify the connection
+			// with the first protocol we've found.
+			name: "mixed protocols",
+			done: make(chan struct{}),
+			clientRun: func(t *testing.T, serverAddr string) {
+				timedContext, cancel := context.WithTimeout(context.Background(), time.Second)
+				c, err := dialer.DialContext(timedContext, "tcp", serverAddr)
+				cancel()
+				require.NoError(t, err)
+				defer c.Close()
+				c.Write([]byte("HTTP/1.1 404 NOT FOUND\r\n"))
+				// http2 prefix.
+				c.Write([]byte("PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n"))
+			},
+			serverRun: func(t *testing.T, serverAddr string, done chan struct{}) {
+				server := NewTCPServerOnAddress(serverAddr, func(c net.Conn) {
+					r := bufio.NewReader(c)
+					input, err := r.ReadBytes(byte('\n'))
+					if err == nil {
+						c.Write(input)
+					}
+					c.Close()
+				})
+				require.NoError(t, server.Run(done))
+			},
+			want: network.ProtocolHTTP,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			tr, err := NewTracer(cfg)
+			if err != nil {
+				t.Fatal(err)
+			}
+			defer tr.Stop()
+			defer close(tt.done)
+
+			initTracerState(t, tr)
+			port, err := GetFreePort()
```

Comment:
I don't think you need to do this. You can just pass in `serverHost + ":0"` to `tt.serverRun` and have that function return the `TCPServer*`. You can then pass `server.address` to `tt.clientRun`.

Reply:
I need the port to create serveradrees and target address
Othewise i need to run the server, expect it to return the binded address and then create the target address

Wdyt?

Reply:
That's what I am suggesting, would that not work? If you have `serverRun` return the `TCPServer*`, the address for the client should be there.

Reply:
Oh I see, you would have to extract the port from the returned address. But I think that is preferable; the `GetFreePort` function does not guarantee the port will be free which could cause flakes in the CI.

Reply:
ill make the changed